types/next/document.d.ts

```diff
@@ -17,13 +38,7 @@ export interface Context {
     };
     asPath: string;
 
-    renderPage(
-        enhancer?: (page: React.Component) => React.ComponentType<any>
-    ): {
-        html?: string;
-        head: Array<React.ReactElement<any>>;
-        errorHtml: string;
-    };
+    renderPage<E extends PageProps = AnyPageProps, P extends any = E>(enhancer?: Enhancer<E, P>): RenderPageResponse; // tslint:disable-line:no-unnecessary-generics
```

Comment:
`tslint` wants me to remove the generic types here, but then `Enhancer` can't receive types, and is set to `AnyPageProps, AnyPageProps`, which may conflict with the actual types in use. I don't know of another way to write this; I think that this is valid. I could, I suppose, _not_ define a separate type for `Enhancer`, and instead put it inline with the argument. But being able to export it seems useful. I disabled the rule for this line as it was suggested I might be allowed to do [here](https://github.com/Microsoft/dtslint/issues/76#issuecomment-364201501) in a discussion of this linter rule.

Reply:
Wait, I don't understand why there's a conflict with the actual types in use. Can you give me an example of what you'd like to be able to write?

```ts
renderPage(enhancer?: Enhancer<AnyPageProps, AnyPageProps>): RenderPageResponse;
```

Reply:
You can see the different examples that all work in the tests in this PR. They won't work any other way that I've found:

```tsx
const enhancerExplicit: Enhancer<PageProps, {}> = (App) => (props) => (<App />);
const enhancerInferred = (App: React.ComponentType<ProcessedInitialProps>) => ({ foo, bar }: PageInitialProps) => (<App fooLength={foo.length} bar={!!bar} />);
const explicitEnhancerRenderResponse = renderPage(enhancerExplicit);
const inferredEnhancerRenderResponse = renderPage(enhancerInferred);
const defaultedTypesRenderResponse = renderPage((App) => (props) => (<App url={props.url} />));
const defaultedTypesExtendedRenderResponse = renderPage((App) => (props) => (<App foo="bar" url={props.url} />));
const explicitTypesRenderResponseOne = renderPage<PageProps, {}>((App) => (props) => (<App />));
const explicitTypesRenderResponseTwo = renderPage<PageInitialProps, ProcessedInitialProps>((App) => ({ foo, bar }) => (<App fooLength={foo.length} bar={!!bar} />));
```

With your example:

```ts
renderPage(enhancer?: Enhancer<AnyPageProps, AnyPageProps>): RenderPageResponse;
```

Four of my six tests will break:

![image](https://user-images.githubusercontent.com/331287/40377115-1617c35e-5dbe-11e8-8a3f-57375a581697.png)

Two of them, of course, are now passing types that are not expected—so fixing that:

![image](https://user-images.githubusercontent.com/331287/40377138-293abb4e-5dbe-11e8-9336-98ae8f49680e.png)

Your example hard-codes the `Enhancer`'s types to `AnyPageProps`, which is both overly strict (it requires `url`, which some props may not have) and overly vague (it allows any other props and will never complain about props mismatch as long as `url` exists). I built `AnyPageProps` because that _allows_ permissiveness while still defaulting to using `url`, but hard-coding to that is not a useful solution. Hard-coding to `any` means you _can't_ be specific about what props the two ComponentTypes expect, even if you want to be (and I do).

Reply:
Could you then encode `AnyPageProps` as the following and still go with what was written?

```ts
interface AnyPageProps extends Partial<PageProps> {
  [x: string]: any;
}
```

Reply:
@DanielRosenwasser is this good to go?

Reply:
Sorry everyone I’m away from home at the moment with limited access to stuff but I just tried out the `Partial` approach and it fails the `PageProps` constraint by making `url` optional, while afaik `url` _will always be passed_.

I don’t feel this solution is overly strict, because no types need to be passed along, but they _can_ be and when they do they work properly.